library/src/scala/main.scala

```diff
@@ -8,6 +8,192 @@
 
 package scala
 
+import collection.mutable
+
+
 /** An annotation that designates a main function
  */
-class main extends scala.annotation.Annotation {}
+class main extends scala.annotation.MainAnnotation:
+  self =>
+  import main._
+
+  protected sealed abstract trait Argument {
+    def name: String
+    def typeName: String
+    def doc: String
+
+    def usage: String
+  }
+  protected class SimpleArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[--$name] <$name>"
+  protected class OptionalArgument[T](val name: String, val typeName: String, val doc: String, val defaultValue: T)
+    extends Argument:
+    override def usage: String = s"[[--$name] <$name>]"
+  protected class VarArgument(val name: String, val typeName: String, val doc: String) extends Argument:
+    override def usage: String = s"[<$name> [<$name> [...]]]"
+
+  override type ArgumentParser[T] = util.CommandLineParser.FromString[T]
+  override type MainResultType = Any
+
+  /** Prints the main function's usage */
+  def usage(commandName: String, args: Seq[Argument]): Unit =
+    def wrappedArgumentUsages(argsUsage: List[String], maxLength: Int): List[String] =
+      def recurse(args: List[String], currentLine: String, acc: Vector[String]): Vector[String] =
+        (args, currentLine) match {
+          case (Nil, "") => acc
+          case (Nil, l) => (acc :+ l)
+          case (arg :: t, "") => recurse(t, arg, acc)
+          case (arg :: t, l) if l.length + 1 + arg.length <= maxLength => recurse(t, s"$l $arg", acc)
+          case (arg :: t, l) => recurse(t, arg, acc :+ l)
+        }
+
+      recurse(argsUsage, "", Vector()).toList
+    end wrappedArgumentUsages
+
+    val maxLineLength = 120 // TODO as parameter? As global Dotty parameter?
```

Comment:
I wonder if this would work
```scala
class main(maxLineLength: Int = 120)
```

```scala
@main def ...
@main(80) def ...
@main(maxLineLength=80) def ...
```